### DISCUSSION TITLE
Local deployment configuration

### DISCUSSION THREAD

**Discussion by multimeric** [wlandau/crew#89] [Q&A]
I've noticed with the change from e.g. `future` to `crew` in targets, we've moved the configuration of the execution engine into `_targets.R` instead of it being passed into `tar_make_future()`. For example, this is true of the number of workers.

I'm therefore wondering how we can implement portable configuration, which can be modified depending on the execution environment. For example a given environment might have more CPUs and could therefore have a higher number of workers than another environment. Can we or should we ever use `tar_option_set` in the R console before using `tar_make()`?

**Comment by wlandau** [comment]
> I've noticed with the change from e.g. future to crew in targets, we've moved the configuration of the execution engine into _targets.R instead of it being passed into tar_make_future(). For example, this is true of the number of workers.

The number of workers is passed to `tar_make_future()`, but everything else about the `future::plan()` happens inside the `_targets.R` file. `crew` moves `workers` into the `_targets.R` file, but otherwise the portability issue is the same as for `future`.

> I'm therefore wondering how we can implement portable configuration, which can be modified depending on the execution environment. For example a given environment might have more CPUs and could therefore have a higher number of workers than another environment. Can we or should we ever use tar_option_set in the R console before using tar_make()?

`tar_make()` is called in your local R session, but then `_targets.R` is sourced inside a new R process that is protected from your local session. In other words, if you set `tar_option_set()` in your local session, those options will not apply to the pipeline.

When moving from machine to machine, you might need to edit the `controller` option in `tar_option_set()` inside `_targets.R`, but the other aspects of the project should not need to change. The only exception I can think of is if you set a complicated controller group and and assign individual targets to individual controllers, and something about a change in platform requires you remove one of those controllers and reassign individual targets to different controllers.

**Comment by brendanf** [comment]
Isn't this the textbook use case for  `_targets.yaml`?

**Reply by multimeric** [reply]
Could you show me an example of how this might be used to configure the number of `crew` workers?

**Reply by brendanf** [reply]
I don't think the `workers` field in _targets.yaml has any hard-coded effect on `crew`, but you can make `_targets.yaml` once per machine via:
```R
tar_config_set(workers = 8) # or whatever value you want on this machine
```
which produces `_targets.yaml`:
```yaml
main:
  workers: '8'
```


Then, in `_targets.R` you can do:
```R
...
tar_option_set(controller = crew_controller_local(workers = tar_config_get("workers"))
...
```

**Reply by wlandau** [reply]
> Isn't this the textbook use case for _targets.yaml?

`_targets.yaml` is for arguments to functions called in the local session, such as `tar_make()` and tar_outdated()`. `tar_option_set()` is for configuration inside `_targets.R`. The former is in your (possibly interactive) parent R session, and the latter runs inside a `callr::r()` process and cannot be accessed interactively.

The workaround in https://github.com/wlandau/crew/discussions/89#discussioncomment-6188063 should be fine in many cases, but I think it is safer the  completely general case to manually update the `workers` argument of the `crew` controller in `_targets.R`.
